android/app/src/main/java/app/organicmaps/MwmActivity.java

```diff
@@ -2148,6 +2245,148 @@ public void onSettingsOptionSelected()
     startActivity(intent);
   }
 
+  private boolean showTracePathDisclaimer()
+  {
+    if (Config.isTracePathDisclaimerAccepted())
+      return true;
+
+    final StringBuilder builder = new StringBuilder();
+    for (int resId : new int[]{ R.string.instruction_1, R.string.instruction_2, R.string.instruction_3, R.string.instruction_4, R.string.have_a_nice_day})
+      builder.append(getString(resId)).append("\n\n");
+
+    dismissAlertDialog();
+    mAlertDialog = new MaterialAlertDialogBuilder(this, R.style.MwmTheme_AlertDialog)
+        .setTitle(R.string.Alert_dialogue_title)
+        .setMessage(builder.toString())
+        .setCancelable(false)
+        .setNegativeButton(R.string.cancel, null)
+        .setPositiveButton(R.string.ok, (dlg, which) -> {
+          Config.acceptTracePathDisclaimer();
+          startTracePath(true);
+        })
+        .setOnDismissListener(dialog -> mAlertDialog = null)
+        .show();
+
+    return false;
+  }
+
+  private void startTracePathSafe(boolean enable)
+  {
+    resetFlags();
+    startTracePath(enable);
+  }
+
+  private void startTracePath(boolean enable)
+  {
+    if(enable)
+    {
+      if(!LocationUtils.checkLocationPermission(this))
+      {
+        Logger.i(TAG,"Location permission not granted");
+        dismissLocationErrorDialog();
+
+        // This variable is a simple hack to re initiate the flow
+        // according to action of user. Calling it hack because we are avoiding
+        // creation of new methods by using this variable.
+        mTracePathLocationPermission = true;
```

Comment:
1. Let's rename TracePath to TrackRecorder everywhere for consistency and to avoid renaming later.
2. The use of this variable is not clear, is there a better way/naming to make it clearer?